pkg/common/controlprotocol/controlprotocol.go

```diff
@@ -0,0 +1,110 @@
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controlprotocol
+
+import (
+	"context"
+
+	ctrl "knative.dev/control-protocol/pkg"
+	"knative.dev/control-protocol/pkg/message"
+	"knative.dev/control-protocol/pkg/network"
+	ctrlservice "knative.dev/control-protocol/pkg/service"
+)
+
+const (
+	// ServerPort is the port used when initializing the control-protocol server
+	ServerPort = 8085
+)
+
+// AsyncHandlerFunc is an alias that matches the type used in the control-protocol's NewAsyncCommandHandler function
+type AsyncHandlerFunc = func(ctx context.Context, commandMessage ctrlservice.AsyncCommandMessage)
+
+// startServerWrapper wraps the Control Protocol initialization call to facilitate
+// unit testing without needing to start live TCP servers
+var startServerWrapper = network.StartInsecureControlServer
+
+// ServerHandler defines the interface for adding and removing sync or async handlers from a
+// control-protocol server
+type ServerHandler interface {
+	Shutdown()
+	AddAsyncHandler(opcode ctrl.OpCode, resultOpcode ctrl.OpCode, payloadType message.AsyncCommand, handler AsyncHandlerFunc)
+	AddSyncHandler(opcode ctrl.OpCode, handler ctrl.MessageHandlerFunc)
+	RemoveHandler(opcode ctrl.OpCode)
+}
+
+// serverHandlerImpl is the primary implementation of a ServerHandler
+type serverHandlerImpl struct {
+	router       ctrlservice.MessageRouter
+	server       *network.ControlServer
+	cancelServer context.CancelFunc
+}
+
+// Verify that the serverHandlerImpl implements the ServerHandler interface
+var _ ServerHandler = (*serverHandlerImpl)(nil)
+
+// NewServerHandler starts a control-protocol server on the specified port and returns
+// the serverHandlerImpl as a ServerHandler interface
+func NewServerHandler(ctx context.Context, port int) (ServerHandler, error) {
+	serverCtx, serverCancelFn := context.WithCancel(ctx)
+
+	// Create a new control-protocol server that will listen on the given port
+	controlServer, err := startServerWrapper(serverCtx, network.WithPort(port))
+
+	if err != nil {
+		serverCancelFn()
+		return nil, err
+	}
+
+	messageRouter := make(ctrlservice.MessageRouter)
+	controlServer.MessageHandler(messageRouter)
+
+	return serverHandlerImpl{
+		server:       controlServer,
+		router:       messageRouter,
+		cancelServer: serverCancelFn,
+	}, nil
+}
+
+// Shutdown cancels the control-protocol server and wait for it to stop
+func (s serverHandlerImpl) Shutdown() {
+	s.cancelServer()
+	waitChannelClosed(s.server.ClosedCh())
+}
```

Comment:
do we need a timeout of some sort here - will it block shutdown forever if channel doesn't return?

Reply:
That's a good idea; it also removes the need for the extra nil channel test